theories/datatypes/List.ec

```diff
@@ -1818,6 +1819,137 @@ lemma mapiP x0 (f : int -> 'a -> 'b) (s : 'a list) y :
+  put (rcons s y) (size s) x = rcons s x.
+proof. 
+rewrite /put size_rcons size_ge0 (_ : size s < size s + 1) 1:/# /=. 
+by rewrite -{1}cats1 drop_oversize 1:size_rcons // take_cat cats0 cats1.
+qed.
+
+lemma put_rcons (s : 'a list) (n : int) (x y : 'a) :
+  n <> size s => put (rcons s y) n x = rcons (put s n x) y.
+proof. 
+move => ?; rewrite -?cats1 put_cat.
+by case (n < size s) => // /lezNgt ?; rewrite ?put_oversize /#.
+qed.
+
+lemma nth_put (dflt : 'a) (s : 'a list) (n : int) (x : 'a) :
+  0 <= n < size s => nth dflt (put s n x) n = x.
+proof.
+by move => [ge0_n lts1_n] @/put; rewrite ge0_n lts1_n /= nth_cat size_take 2:lts1_n.
+qed.
+
+lemma nth_put_out (dflt : 'a) (s : 'a list) (n : int)  (x : 'a) :
+  !(0 <= n < size s) => nth dflt (put s n x) n = dflt.
+proof. by move => ?; rewrite put_out 2:nth_out. qed.
+
+lemma nth_put_neq (dflt : 'a) (s : 'a list) (n1 n2 : int) (x : 'a) :
+  n1 <> n2 => nth dflt (put s n1 x) n2 = nth dflt s n2.
+proof.
+move => ? @/put; case (0 <= n1 < size s) => [[? lts_n1]|] //.
+rewrite nth_cat size_take 2:lts_n1 //= ; case (n2 < n1) => ?.
++ by rewrite nth_take.
++ by rewrite (: n2 - n1 <> 0) /= 2:nth_drop /#.
+qed.
+
+lemma eq_after_put (dflt : 'a) (s1 s2 : 'a list) (n : int) :
+     size s1 = size s2
+  => (forall (i : int), i <> n => nth dflt s1 i = nth dflt s2 i)
+  => s1 = put s2 n (nth dflt s1 n).
+proof.
+move => eq_size eq_nth. 
+apply (eq_from_nth dflt s1 (put s2 n (nth dflt s1 n))) => [| i rng_i]. 
++ by rewrite size_put.
++ case (i <> n) => [neqn_i | /= eqn_i].
+  - move: (eq_nth i); rewrite neqn_i => -> //.
+    by rewrite nth_put_neq 1:eq_sym.
+  - by rewrite eqn_i nth_put /#.
+qed.
+
+lemma mem_put (s : 'a list) (n : int)  (x : 'a) :
+  0 <= n < size s => mem (put s n x) x.
+proof.
+by move => rng_n; rewrite /put rng_n /= mem_cat; right; rewrite in_cons.
+qed.
+
+lemma map_put (f : 'a -> 'b) (s : 'a list) (n : int) (x : 'a) :
+  map f (put s n x) = put (map f s) n (f x).
+proof.
+case (0 <= n < size s) => rng_n @/put; rewrite size_map rng_n //=.
+by rewrite map_cat map_cons map_take map_drop.
+qed.
+
+lemma nth_map_put (dflt : 'b) (f : 'a -> 'b) (s : 'a list) (n : int) (x : 'a) :
+  0 <= n < size s => nth dflt (map f (put s n x)) n = f x.
+proof. by move => ?; rewrite map_put nth_put 1:size_map. qed.
+
+lemma nth_map_put_out (dflt : 'b) (f : 'a -> 'b) (s : 'a list) (n : int) (x : 'a) :
+  !(0 <= n < size s) => nth dflt (map f (put s n x)) n = dflt.
+proof. by move => ?; rewrite map_put nth_put_out 1:size_map. qed.
+
+lemma eq_nth_put (dflt1 dflt2 : 'a) (s1 s2 : 'a list) (n1 n2 : int) (x : 'a) :
+  0 <= n1 < size s1 => 0 <= n2 < size s2 =>
+  nth dflt1 (put s1 n1 x) n1 = nth dflt2 (put s2 n2 x) n2.
+proof. by move => *; rewrite ?nth_put. qed.
+
+lemma eq_nth_put_out (dflt : 'a) (s1 s2 : 'a list) (n1 n2 : int) (x1 x2 : 'a) :
+  !(0 <= n1 < size s1) => !(0 <= n2 < size s2) =>
+  nth dflt (put s1 n1 x1) n1 = nth dflt (put s2 n2 x2) n2.
+proof. by move => *; rewrite ?nth_put_out. qed.
```

Comment:
Not sure these are needed, as they seem to be simple combinations of existing lemmas.